### ISSUE TITLE
[Bug]: macOS password prompt when using safeStorage after electron upgrade

### ISSUE THREAD

**Issue by lunaaa3** [electron/electron#43233]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a bug report that matches the one I want to file, without success.

### Electron Version

31.3.1

### What operating system(s) are you using?

macOS

### Operating System Version

Sonoma 14.5

### What arch are you using?

arm64 (including Apple Silicon)

### Expected Behavior

After upgrading electron, safeStorage can still be used without showing the macOS password prompt.

### Actual Behavior

I ran into the exact same issues as described in https://github.com/electron/electron/issues/40236 (as this issue was closed, I opened this one).

### Additional Information

**Steps to reproduce:**
1. Start below example (npm start) --> Encryption works and buffer printed in console
2. Upgrade electron (npm install electron@31.3.1 --save-dev)
3. Run example again after upgrade (npm start) --> **macOS password prompt shown and encryption/decryption doesn't work until password entered.** 

**package.json**
```javascript
{
  "name": "electron-safestorage-upgrade-bug",
  "main": "main.js",
  "scripts": {
    "start": "electron ."
  },
  "devDependencies": {
    "electron": "30.3.1"
  }
}
``` 

**main.js**
```javascript
const { app, safeStorage } = require("electron/main");

app.whenReady().then(() => {
  useSafeStorage();
});

const useSafeStorage = async () => {
  try {
    const encryptedString = await safeStorage.encryptString("hello world");
    console.log(encryptedString);
  } catch (e) {
    console.log(e);
  }
};
```

**Comment by nmggithub** [comment]
Is there a screenshot of the specific password prompt? That would be very helpful in isolating the issue, as there are many things that prompt a password in macOS.

**Comment by nikwen** [comment]
Question: Does this only happen in development or also when you package, codesign and notarize the app?

I've seen similar issues with other macOS APIs in development (e.g. `app.setLoginItemSettings()`), but everything worked once the app was packaged, codesigned and notarized.

I assume macOS might not be able to deduce that an unsigned Electron executable and another unsigned Electron executable are the same app. In production, however, when everything is codesigned and notarized, it is able to.

**Comment by nmggithub** [comment]
> Question: Does this only happen in development or also when you package, codesign and notarize the app?
> 
> I've seen similar issues with other macOS APIs in development (e.g. `app.setLoginItemSettings()`), but everything worked once the app was packaged, codesigned and notarized.
> 
> I assume macOS might not be able to deduce that an unsigned Electron executable and another unsigned Electron executable are the same app. In production, however, when everything is codesigned and notarized, it is able to.

I definitely would not be surprised if this was the case. I've looked into macOS keychain stuff when doing work unrelated to Electron and I've seen that each keychain item can have it's own access-control list. The way Electron seems to handle in the current version (v32.0.1) it is to have [a single keychain item named after the app](https://github.com/electron/electron/blob/v32.0.1/shell/browser/electron_browser_main_parts.cc#L541). I believe it's been this way for a while. I'm currently digging through the Chromium source to see if and how it sets the access-control lists for these items.

**Comment by nmggithub** [comment]
OK. I've figured it out mostly. Here is the breakdown.

_Note: Version numbers mentioned are the most recent at the time of writing. Discussion of behavior on current versions does not preclude the existence of the same behavior on previous versions._

**Electron Version**: v32.0.1
**Chromium Version**: [v128.0.6613.36](https://source.chromium.org/chromium/chromium/src/+/refs/tags/128.0.6613.36:)
Source: https://releases.electronjs.org/release/v32.0.1

**macOS Version**: 14.6
**Security framework verson**: [61123.140.15 (commit: b4bd4fda62af1ebf469ee97f0112fb49363fb482)](https://github.com/apple-oss-distributions/Security/tree/b4bd4fda62af1ebf469ee97f0112fb49363fb482)
Source: https://github.com/apple-oss-distributions/distribution-macOS/tree/macos-146

# Electron/Chromium implementation of `safeStorage.encryptString()`

## [`EncryptString`](https://github.com/electron/electron/blob/v32.0.1/shell/browser/api/electron_api_safe_storage.cc#L50)

Calling `safeStorage.encryptString` in JS ends up calling this C++ code, which calls `OSCrypt::EncryptString`.

## [`OSCrypt::EncryptString`](https://source.chromium.org/chromium/chromium/src/+/refs/tags/128.0.6613.36:components/os_crypt/sync/os_crypt_mac.mm;l=55)
`OSCrypt::EncryptString` in Chromium is just a wrapper for `OSCryptImpl::GetInstance()->EncryptString`.

## [`OSCryptImpl::EncryptString`](https://source.chromium.org/chromium/chromium/src/+/refs/tags/128.0.6613.36:components/os_crypt/sync/os_crypt_mac.mm;l=157)

`OSCryptImpl::EncryptString` calls `GetEncryptionKey`.

## [`OSCryptImpl::GetEncryptionKey`](https://source.chromium.org/chromium/chromium/src/+/refs/tags/128.0.6613.36:components/os_crypt/sync/os_crypt_mac.mm;l=90)

`OSCryptImpl::GetEncryptionKey` calls `KeychainPassword::GetPassword()`

## [`KeychainPassword::GetPassword()`](https://source.chromium.org/chromium/chromium/src/+/refs/tags/128.0.6613.36:components/os_crypt/sync/keychain_password_mac.mm;l=81)

`KeychainPassword::GetPassword()` attempts to get the password stored in the keychain via the service name and account name. If the account does not exist, it creates one via `AddRandomPasswordToKeychain`. Note that for macOS builds of Electron, [the service name and account name are assigned/overwritten to be named after the given Electron app at build time](https://github.com/electron/electron/blob/v32.0.1/shell/browser/electron_browser_main_parts.cc#L540-L543)

## [`AddRandomPasswordToKeychain`](https://source.chromium.org/chromium/chromium/src/+/refs/tags/128.0.6613.36:components/os_crypt/sync/keychain_password_mac.mm;l=41)

`AddRandomPasswordToKeychain` creates a random password, and then calls `keychain.AddGenericPassword`.

## [`keychain.AddGenericPassword`](https://source.chromium.org/chromium/chromium/src/+/refs/tags/128.0.6613.36:crypto/apple_keychain_mac.cc;l=75)
`keychain.AddGenericPassword` calls `SecKeychainAddGenericPassword`.

## [`SecKeychainAddGenericPassword` documentation](https://developer.apple.com/documentation/security/seckeychainaddgenericpassword(_:_:_:_:_:_:_:_:))

`SecKeychainAddGenericPassword` is function that is part of the Security framework on macOS. In the documentation, it states:

> This function sets the initial access rights for the new keychain item so that the application creating the item is given trusted access.

## [`SecKeychainAddGenericPassword` source](https://github.com/apple-oss-distributions/Security/blob/b4bd4fda62af1ebf469ee97f0112fb49363fb482/OSX/libsecurity_keychain/lib/SecKeychain.cpp#L766)

After looking at the source code for the `SecKeychainAddGenericPassword` function, it does not appear to explicitly do what Apple's documentation claims it does. However, it's possible this is done elsewhere in the code, or in sections that are not open-sourced by Apple.

# In short

On macOS: 
1. Electron's `safeStorage` API makes use of
2. Chromium's `OSCrypt` API, which makes use of
3. macOS's `SecKeychain` API, to setup a single keychain item to be used across `OSCrypt` calls, and
4. macOS's `SecKeychain` API automatically limits access to the item so that only the creating application is given trusted access (password-less access).

So whatever indicators that macOS uses to confirm if the application that is accessing the keychain item is, indeed, the same one that created it initially do not survive rebuilds (or potentially even subsequent runs). As @nikwen mentioned, it's very likely related to the code signatures.

**Comment by hipstersmoothie** [comment]
We're hitting this too and have been forever. We recently rewrote our app to be basically a shell, updated all our electron deps, and even switched from builder to forge and we still see the issue.

It happens mostly in prod and mostly to customers. If a person encounters this they basically have to constantly accept the prompt.

**Comment by nmggithub** [comment]
> We're hitting this too and have been forever. We recently rewrote our app to be basically a shell, updated all our electron deps, and even switched from builder to forge and we still see the issue.
> 
> It happens mostly in prod and mostly to customers. If a person encounters this they basically have to constantly accept the prompt.

See my breakdown above. This isn't really a bug in Electron. It's just macOS not understanding that the application that is attempting to access that keychain item is the same one that created it. Again, I'm not sure why exactly it does that, but I don't think it's an Electron issue.

To debug it, you can find a macOS machine it's happening on and go into the Access Control section like I mentioned above. Except instead of changing it to allow all applications, take a look at what the access control settings are. Your application itself should be in the list of applications to access it. If it isn't, that's weird. If you see it there, but the password prompt is still happening, try removing it and re-adding it. There's some additional metadata that Apple doesn't really let you change directly (without coding something yourself), but removing re-adding the application should provide it with the latest code signature of your application.

I'm willing to provide additional help if needed.

**Comment by MarshallOfSound** [comment]
This won't happen if your app is codesigned correctly, the ACL is based on the app bundle ID + the team ID of the certificate used to sign your app (by default). Under the hood it's actually the DR used to signed your app but by default that translates to bundle ID + team ID.

If you aren't signing your app this is going to happen constantly, if people are re-signing your app it's going to happen constantly. If you sign your app with multiple teams certs / have migrated between apple teams then this will happen constantly. But just don't do that and sign your app properly and you'll be fine 😅

**Comment by nmggithub** [comment]
> This won't happen if your app is codesigned correctly, the ACL is based on the app bundle ID + the team ID of the certificate used to sign your app (by default). Under the hood it's actually the DR used to signed your app but by default that translates to bundle ID + team ID.
> 
> If you aren't signing your app this is going to happen constantly, if people are re-signing your app it's going to happen constantly. If you sign your app with multiple teams certs / have migrated between apple teams then this will happen constantly. But just don't do that and sign your app properly and you'll be fine 😅 


+1 to this. I can definitely confirm on my end, having looked into macOS keychain ACL's for unrelated reasons (even before my deep dive above), this is all correct. By the way, for anyone confused, DR stands for "designated requirement".

https://developer.apple.com/documentation/technotes/tn3127-inside-code-signing-requirements#Designated-requirement

**Comment by nikwen** [comment]
On a related note, #40430 could be useful here. Use different keychain profiles in development and production.

**Comment by adadgio** [comment]
It's often because both apps have a different bundleID (or name?) but attempt to access the same keychain Safe storage entry. Even if you don't use it in your app, some bundles might still do.

Haven't yet found a solution.

**Comment by nmggithub** [comment]
> It's often because both apps have a different bundleID (or name?) but attempt to access the same keychain Safe storage entry. Even if you don't use it in your app, some bundles might still do.
> 
> Haven't yet found a solution.

If they have different bundle ID's, they are different apps. They should not (in both sense of the phrase) be using the same keychain item if they are different apps.

**Comment by adadgio** [comment]
> > It's often because both apps have a different bundleID (or name?) but attempt to access the same keychain Safe storage entry. Even if you don't use it in your app, some bundles might still do.
> > Haven't yet found a solution.
> 
> If they have different bundle ID's, they are different apps. They should not (in both sense of the phrase) be using the same keychain item if they are different apps.

I just tested this with a proper A/B est, and definitely not. The relevant parameter is the app name (as per `app.setName("xxx")`. Apps with the same name BUT different bundle IDs still try to access the same Safe Storage item in keychain. 

Which also means that if you create a beta and prod build for instance, but both have the same name, you'll get the pop-up on which ever app was installed after, even with different bundle ids.

**Comment by nmggithub** [comment]
> > > It's often because both apps have a different bundleID (or name?) but attempt to access the same keychain Safe storage entry. Even if you don't use it in your app, some bundles might still do.
> > > Haven't yet found a solution.
> > 
> > 
> > If they have different bundle ID's, they are different apps. They should not (in both sense of the phrase) be using the same keychain item if they are different apps.
> 
> I just tested this with a proper A/B est, and definitely not. The relevant parameter is the app name (as per `app.setName("xxx")`. Apps with the same name BUT different bundle IDs still try to access the same Safe Storage item in keychain.
> 
> Which also means that if you create a beta and prod build for instance, but both have the same name, you'll get the pop-up on which ever app was installed after, even with different bundle ids.

I guess that makes sense to some extent (as the keychain item itself is named after the app name). In any case, I do believe two different apps cannot use the same keychain item.

**Comment by adadgio** [comment]
I guess that depends what "two different apps means" then.
